core/lib/src/fs/server.rs

```diff
@@ -60,351 +72,492 @@ use crate::fs::NamedFile;
+#[derive(Debug)]
+pub enum FileResponse<'p, 'h> {
+    /// Return the contents of the specified file.
+    File(File<'p, 'h>),
+    /// Returns a Redirect to the specified path. This needs to be an absolute
+    /// URI, so you should start with [`File.full_uri`](File) when constructing
+    /// a redirect.
+    Redirect(Redirect),
+}
 
-        FileServer { root: path.into(), options, rank: Self::DEFAULT_RANK }
+impl<'p, 'h> From<File<'p, 'h>> for FileResponse<'p, 'h> {
+    fn from(value: File<'p, 'h>) -> Self {
+        Self::File(value)
     }
+}
+impl<'p, 'h> From<File<'p, 'h>> for Option<FileResponse<'p, 'h>> {
+    fn from(value: File<'p, 'h>) -> Self {
+        Some(FileResponse::File(value))
+    }
+}
 
-    /// Sets the rank for generated routes to `rank`.
-    ///
-    /// # Example
-    ///
-    /// ```rust,no_run
-    /// use rocket::fs::{FileServer, Options};
-    ///
-    /// // A `FileServer` created with `from()` with routes of rank `3`.
-    /// FileServer::from("/public").rank(3);
-    ///
-    /// // A `FileServer` created with `new()` with routes of rank `-15`.
-    /// FileServer::new("/public", Options::Index).rank(-15);
-    /// ```
-    pub fn rank(mut self, rank: isize) -> Self {
-        self.rank = rank;
+impl<'p, 'h> From<Redirect> for FileResponse<'p, 'h> {
+    fn from(value: Redirect) -> Self {
+        Self::Redirect(value)
+    }
+}
+impl<'p, 'h> From<Redirect> for Option<FileResponse<'p, 'h>> {
+    fn from(value: Redirect) -> Self {
+        Some(FileResponse::Redirect(value))
+    }
+}
+
+/// A File response from a [`FileServer`]
+#[derive(Debug)]
+pub struct File<'p, 'h> {
+    /// The path to the file that [`FileServer`] will respond with.
+    pub path: Cow<'p, Path>,
+    /// The original uri from the [`Request`]
+    pub full_uri: &'p Origin<'p>,
+    /// A list of headers to be added to the generated response.
+    pub headers: HeaderMap<'h>,
+}
+
+impl<'p, 'h> File<'p, 'h> {
+    /// Add a header to this `File`.
+    pub fn with_header<'n: 'h, H: Into<Header<'n>>>(mut self, header: H) -> Self {
+        self.headers.add(header);
         self
     }
+
+    /// Replace the path of this `File`.
+    pub fn with_path(self, path: impl Into<Cow<'p, Path>>) -> Self {
+        Self {
+            path: path.into(),
+            full_uri: self.full_uri,
+            headers: self.headers,
+        }
+    }
+
+    /// Replace the path of this `File`.
+    pub fn map_path<R: Into<Cow<'p, Path>>>(self, f: impl FnOnce(Cow<'p, Path>) -> R) -> Self {
+        Self {
+            path: f(self.path).into(),
+            full_uri: self.full_uri,
+            headers: self.headers,
+        }
+    }
+
+    /// Convert this `File` into a Redirect, transforming the URI.
+    pub fn into_redirect(self, f: impl FnOnce(Origin<'static>) -> Origin<'static>)
+        -> FileResponse<'p, 'h>
+    {
+        FileResponse::Redirect(Redirect::permanent(f(self.full_uri.clone().into_owned())))
+    }
```

Comment:
This feels wrong to me. Remove it falls in line with my comments on `full_uri` and the `normalize_dirs` rewriter.